### PR TITLE
Avoid test-only logic in prod code by using Local Nuget Source to retrieve baseline

### PR DESCRIPTION
**Pull request by RodgeFu**
Avoid test-only logic in prod code by using Local Nuget Source to retrieve baseline

### CODE REVIEW

File: src/AutoRest.CSharp/Common/Input/Source/SourceInputModel.cs [hunk 1]
```diff
@@ -159,16 +159,12 @@ private bool TryGetName(ISymbol symbol, [NotNullWhen(true)] out string? name)
                 {
                     return await GeneratedCodeWorkspace.CreatePreviousContractFromDll(Path.Combine(nugetFolder, $"{Configuration.Namespace}.xml"), fullPath);
                 }
+                else
+                {
+                    throw new InvalidOperationException($"Can't find Baseline contract assembly ({baselineVersion}) from Nuget Global Package Folder at {fullPath}. " +
```
Comment (live1206) [inline src/AutoRest.CSharp/Common/Input/Source/SourceInputModel.cs]: Consider to display the package name as well.

File: test/NugetPackages/.gitignore [hunk 2]
```diff
@@ -0,0 +1 @@
+!*.nupkg
```
Comment (live1206) [inline test/NugetPackages/.gitignore:1]: Does git ignore `.nupkg` by default?
If this is needed, consider to add this to the global `.gitignore`.

Or, we can have source code in `BaselineContract` and do `dotnet pack` on the fly in the script instead.

File: test/TestProjects/MgmtCustomizations/readme.md [hunk 3]
```diff
@@ -12,6 +12,7 @@ namespace: MgmtCustomizations
 model-namespace: false
 public-clients: false
 head-as-boolean: false
+skip-csproj: true
```
Comment (live1206) [inline test/TestProjects/MgmtCustomizations/readme.md:15]: why is this needed?

File: test/scripts/LocalTestNugetSource.psm1 [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      <TargetFramework>netstandard2.0</TargetFramework>
+      <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+      <Nullable>annotations</Nullable>
+    </PropertyGroup>
+  </Project>' | Out-File $projFile
+  "Nuget ProjectFolder created at $projFile" | Out-Host
+
+  dotnet nuget add source $SourceFolder --name $nugetSourceName
+
+  dotnet nuget push $packageFolder\MgmtCustomizations.1.0.0-beta.20240102.1.nupkg -s $nugetSourceName
```
Comment (live1206) [inline test/scripts/LocalTestNugetSource.psm1]: consider to remove the beta version part, make it `1.0.0`

File: eng/Generate.ps1 [hunk 5]
```diff
@@ -382,6 +383,13 @@ if (!$noBuild) {
     Invoke-TypeSpecSetup
 }
 
+$localTestNugetSourceCreated = $false
+$localTestNugetWorkFolder = Join-Path ([System.IO.Path]::GetTempPath()) ([string][System.Guid]::NewGuid())
+# Only MgmtCustomizations project needs the LocalNugetServer now
+if($keys -contains "MgmtCustomizations"){
```
Comment (live1206) [inline eng/Generate.ps1]: How about keeping `BaselineContract` folder and put the corresponding `.nupkg` files under the corresponding folder by name convention?
And then, we can check if the folder and package file exists for the current test project instead of hard-coded name.

File: test/scripts/CleanupLocalTestNugetSource.ps1 [hunk 6]
```diff
@@ -0,0 +1,10 @@
+param(
+    [Parameter(HelpMessage="The folder contains created Source and temp project to clean up if given")]
+    [string] $workFolder,
+
+    [Parameter(HelpMessage="The Name of the Nuget Source to cleanup(remove)")]
+    [string] $nugetSourceName = "localTestNugetSource")
+    
+Import-Module "$PSScriptRoot\LocalTestNugetSource.psm1" -DisableNameChecking -Force;
+
+Remove-Local-Test-Nuget-Source $workFolder $nugetSourceName
```
Comment (live1206) [inline test/scripts/CleanupLocalTestNugetSource.ps1]: This is not in use, can be removed.

File: test/scripts/PrepareLocalTestNugetSource.ps1 [hunk 7]
```diff
@@ -0,0 +1,11 @@
+param(
```
Comment (live1206) [inline test/scripts/PrepareLocalTestNugetSource.ps1]: This is not in use, can be removed

File: test/scripts/LocalTestNugetSource.psm1 [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+  }
+}
+
+function New-Local-Test-Nuget-Source([string]$workFolder, [string]$nugetSourceName="localTestNugetSource"){
+
+  New-Directory-If-Not-Exist $workFolder;
+  $packageFolder = Join-Path $PSScriptRoot "..\NugetPackages"
+  $sourceFolder = Join-Path $workFolder "Source"
+  New-Directory-If-Not-Exist $sourceFolder
+  "Nuget SourceFolder created at $sourceFolder" | Out-Host
```
Comment (live1206) [inline test/scripts/LocalTestNugetSource.psm1]: Is it possible to use the existing folder as `sourceFolder` directly?

File: test/scripts/LocalTestNugetSource.psm1 [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+  $projFile = Join-Path $projFolder "LocalNuget.csproj"
+  '<Project Sdk="Microsoft.NET.Sdk">
+    <!-- This is a dummy project created by script to trigger nuget package install -->
+    <PropertyGroup>
+      <TargetFramework>netstandard2.0</TargetFramework>
+      <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+      <Nullable>annotations</Nullable>
+    </PropertyGroup>
+  </Project>' | Out-File $projFile
+  "Nuget ProjectFolder created at $projFile" | Out-Host
```
Comment (live1206) [inline test/scripts/LocalTestNugetSource.psm1]: It's neat to create the dummy project on the fly and clean it up afterwards!

File: eng/Generate.ps1 [hunk 10]
```diff
@@ -382,6 +383,22 @@ if (!$noBuild) {
     Invoke-TypeSpecSetup
 }
 
+[hashtable]$testPackagesToInstall = @{};
+$localTestNugetSourceFolder = Join-Path $repoRoot 'test/NugetPackages'
+foreach($key in $keys){
+    foreach($nugetPackageFilename in (Get-ChildItem -Path $localTestNugetSourceFolder | Select-Object -ExpandProperty Name)){
+        $nameForRegex = [Regex]::Escape($key);
```
Comment (live1206) [inline eng/Generate.ps1:390]: in which case will we need the escape here, can we just use `$key`? And `$name` could just be `$key`?

Comment (RodgeFu) [reply eng/Generate.ps1:390]: when the key contains special char needs escape for regular expression like '.'. $key should equal to $name. adding var $name and $version just avoids using match group with magic number directly which is hard to read/understand, thx

**Review by live1206 — APPROVED** [review]
